Notifies user for invalid fields while validating

shashigharti opened the pull request:
* Added check for invalid fields: missingValues and fields (in resource descriptor)
* Added check for invalid field: missingValues (in package descriptor)
* Added tests
* Fixed old tests

- fixes #993

shashigharti committed:
notify user for invalid fields

* Added check for invalid fields: missingValues and fields (in resource descriptor)
* Added check for invalid field: missingValues (in package descriptor)
* Added tests
* Fixed old tests

shashigharti committed:
Merge branch 'main' into 993/provides-validation-hints-for-common-mistakes

shashigharti committed:
fixed failed tests

shashigharti commented:
@roll This is ready for the review.

I also tried adding check for the following in package.json. It quietly passes the validation but doesn't trigger metadata_validate function. What would be your suggestion to catch this error?

```
{
  "name": "test-tabulator",
  "fields": [
          {
            "name": "id",
            "type": "integer",            
          },
          {
            "name": "name",
            "type": "string"
          }
        ]
}
```
![screenshot-localhost_8888-2022 05 31-17_23_34](https://user-images.githubusercontent.com/5582809/171164863-eac3e253-3947-411d-a38e-6e694b701177.png)


Thank you!

roll commented on frictionless/package/package.py:
```diff
@@ -704,6 +704,13 @@ def metadata_process(self):
+            if invalid_field in self:
+                note = f"{invalid_field} should be part of {resource} not package."
+                error = errors.ResourceError(note=note)
+                raise FrictionlessException(error)
```
You need to `yield` errors here as in other places in `metadata_validate` functions (the same for resource)

roll replied:
You can write a test for it:

```
package = Package('with-bad-props')
package.metadata_errors == [...this error...]
```

roll commented on tests/package/validate/test_general.py:
```diff
@@ -345,3 +345,14 @@ def test_validate_package_with_resource_data_is_a_string_issue_977():
+    assert report.flatten(["code", "message"]) == [
+        [
+            "resource-error",
+            "The data resource has an error: missingValues should be part of schema/fields not package.",
```
"missingValues" should be set as "resource.schema.missingValues" (not "resource.missingValues")

roll commented on tests/resource/validate/test_general.py:
```diff
@@ -519,3 +519,25 @@ def test_validate_resource_duplicate_labels_with_sync_schema_issue_910():
+    assert report.flatten(["code", "message"]) == [
+        [
+            "resource-error",
+            "The data resource has an error: fields should be part of schema not resource.",
```
"fields" should be set as "resource.schema.fields" (not "resource.fields")

roll commented on frictionless/package/package.py:
```diff
@@ -704,6 +704,13 @@ def metadata_process(self):
 
     def metadata_validate(self):
+        # Check invalid properties
+        invalid_fields = {"missingValues": "schema/fields"}
```
Regarding your question - shall we check for `fields` here also to catch `package.fields`?

roll reviewed (changes requested):
Thanks @shashigharti!

I've added a few comments

shashigharti committed:
fixed tests, metadata error handling

* Replaced exception with yield error messages
* Fixed tests
* renamed test file name

shashigharti committed:
removed irrelevant change

shashigharti committed:
changed test function name

shashigharti commented:
Thank you @roll!.  I have made the changes and it is ready for the review.

roll commented on tests/package/validate/test_general.py:
```diff
@@ -345,3 +345,33 @@ def test_validate_package_with_resource_data_is_a_string_issue_977():
+
+def test_validate_package_with_fields_993():
+    package = Package(descriptor="data/package-with-fields-993.json")
+    output = list(
```
Can we just write assertions like this:
```
assert package.metadata_errors[0].code == ...
assert package.metadata_errors[0].note == ...
```
Your version is smart but really makes the reader think while the simpler one doesn't require any additional energy

shashigharti replied:
Thank you for pointing out. I was making it complex unnecessarily :).

roll commented on tests/package/validate/test_general.py:
```diff
@@ -345,3 +345,33 @@ def test_validate_package_with_resource_data_is_a_string_issue_977():
     ]
+
+
+def test_validate_package_with_missing_values_993():
```
I would add also `package.validate` tests

roll commented on tests/resource/validate/test_general.py:
```diff
@@ -519,3 +519,25 @@ def test_validate_resource_duplicate_labels_with_sync_schema_issue_910():
     ]
+
+
+def test_validate_resource_with_missing_values_993():
```
I would also add `resource.metadata_errors` tests (so it will be 4 for package and 4 for resource)

roll reviewed (changes requested):
@shashigharti Nice!

Just a few minor suggestions

shashigharti committed:
added tests and fixed error messages

1. Added new tests to cover both validate and metadata errors
2. Fixed error message for package by replacing ResourceError with PackageError

shashigharti commented:
@roll  Thank you a lot! I have made the changes.

roll reviewed (approved):
Perfect!
